### PR TITLE
2294 support char in extraction

### PR DESCRIPTION
**Pull request by hiker**
First part of #1990 (kernel extraction with external symbols). Note that some files that are actually processed jinja files are added, since otherwise certain tests/examples of driver kernel extraction do not work (since the .f90 files are inlined in the kernel extraction the source code is required). Alternatively we would have to pre-process these files for CI, this seems to be the easier method.
There is no code/test/example that uses the character type, but support of chars is required to extract from LFRic.

### CODE REVIEW

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for breaking this out Joerg.
All looks fine (modulo some slight tweaks) and I could build the libs locally.
I'm not comfortable with having functionality in the repo. that isn't exercised in some form (although this is also true of the profiling wrappers to a lesser extent). Is it possible to extend one or more examples to exercise this new support for `character` and `long`?
The section 'Jinja support in the base class' in the Dev. Guide needs updating to mention that process.py now supports char too. In fact, so does the 'help' text for the `-types` argument to that script (L61-62).

File: lib/extract/standalone/read_kernel_data_mod.jinja [hunk 1]
```diff
@@ -222,8 +224,12 @@ contains
         {% if name == "Logical" %}
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/read_kernel_data_mod.jinja:220]: Done. I also did this in the NetCDF equivalent code.

File: lib/extract/standalone/read_kernel_data_mod.f90 [hunk 2]
```diff
@@ -143,6 +163,216 @@ subroutine OpenRead(this, module_name, region_name)
...
+        ! Allocate enough space to store the values to be read:
+        allocate(value(dim_size1,dim_size2,dim_size3,dim_size4), Stat=ierr)
+        if (ierr /= 0) then
+            write(stderr,*) "Cannot allocate array for ", name, &
+                            " of size ", dim_size1,dim_size2,dim_size3,dim_size4, &
+                            " in ReadArray4dChar."
+            stop
+        endif
+
+        ! Initialise it with 0, so that an array comparison will work
```
[the remark this answers is not in the document]
Comment (hiker) [reply lib/extract/standalone/read_kernel_data_mod.f90:366]: Agreed.

**Comment by hiker** [comment]
> I'm not comfortable with having functionality in the repo. that isn't exercised in some form (although this is also true of the profiling wrappers to a lesser extent). Is it possible to extend one or more examples to exercise this new support for `character` and `long`? 

Hmm - not easily. Long would be used if the we would pre-process constants_mod.F90 to switch to 64 bit ints, but that's a change I am not keen to make. And I believe no kernel supports character arguments.

I suggest we keep this for #1990 - then I just need to declare a character and long etc in an external module, and that would then be tested.


> The section 'Jinja support in the base class' in the Dev. Guide needs updating to mention that process.py now supports char too. In fact, so does the 'help' text for the `-types` argument to that script (L61-62).

All done.

**Review by arporter — COMMENTED** [review]
All looks nice now - thanks for improving all the comments.
Just waiting on the integration tests now.

**Review by arporter — APPROVED** [review]
Integration tests are green.
Dev. guide builds fine.
Will proceed to merge.

**Pull request merged**